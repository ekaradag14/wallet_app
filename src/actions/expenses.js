import { v4 as uuidv4 } from "uuid";
//Aqui desestructamos el primero objeto. Si tenemos ninguna tenemos un valor defecto. (Un objeto vacio)
export const addExpense = ({
  description = "",
  note = "",
  amount = 0,
  createdAt = 0
} = {}) => ({
  type: "ADD_EXPENSE",
  expense: {
    id: uuidv4(),
    description,
    note,
    amount,
    createdAt
  }
});
//Estas funciones son solo algunos atajos para no escribir tipo cada vez escribimos las funciones
export const removeExpense = ({ id = "" } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});
