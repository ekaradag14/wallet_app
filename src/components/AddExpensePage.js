import React from "react";
import ExpenseModifyComponent from "./ExpenseModifyComponent";
export const AddExpensePage = (props) => {
  return (
    <div>
      <ExpenseModifyComponent
        processCompleted={() => {
          //Una vez acabamos modificando de expense redireccionamos el usuario a componente de tablero
          props.history.push("/");
        }}
      />
    </div>
  );
};

// El primer argumento es null porque no necesitamos alcanzar al store aqui.
export default AddExpensePage;
