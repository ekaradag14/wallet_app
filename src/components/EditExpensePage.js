import React from "react";
import ExpenseModifyComponent from "./ExpenseModifyComponent";
import { connect } from "react-redux";

const EditExpensePage = ({ expenses, match, expense,history }) => {
  return (
    <div>
      {expense && <h3>Editing {expense.description} expense </h3>}
      <ExpenseModifyComponent
        expense={expense}
        processCompleted={() => {
          history.push("/");
        }}
      />
    </div>
  );
};

const mapsStateToProps = (state, props) => { // Pasamos el estado para buscar por el expense 
  //Y Después pasamos ese expense a props para usarlo
  return {
    expense: state.expenses.find((item) => item.id === props.match.params.id)
      //match.params.id es id de nuestra expense la url de pagina es dinámica y construimos con la id
  };
};

export default connect(mapsStateToProps)(EditExpensePage);
