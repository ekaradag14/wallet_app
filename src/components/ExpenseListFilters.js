import "react-dates/lib/css/_datepicker.css";
import React, { useState } from "react";
import {
  setTextFilter,
  sortByDate,
  setStartDate,
  setEndDate,
  sortByAmount
} from "../actions/filters";
import { connect } from "react-redux";
import { DayPickerRangeController } from "react-dates";
import moment from "moment";
const ExpenseListFilters = ({
  putTextFilter,
  putSortDateFilter,
  putAmountFilter,
  putStartDateFilter,
  putEndDateFilter,
  filterValue
}) => {
  const [expenseFilterController, setExpenseFilterController] = useState({
    isInputNecessary: false,
    isCalendarNecessary: false,
    focusedInput: "startDate",
    startDate: null,
    endDate: null
  });
  // Para controlar el estado complejo
  const handleChange = (input) => {
    const { value, name } = input;

    setExpenseFilterController((prev) => {
      return {
        ...prev,
        [name]: value,
      };
    });
  };

  const resetFilters = () => {
    //Antes de ordenar nuestra lista. Necesitamos limpiar nuestra filtros.
    //Tomamos nuestra lista directamente despuÃ©s de filtrarlo
    putStartDateFilter(null);
    putEndDateFilter(null);
    putTextFilter("");
  };

  const handleExtraComponents = (calendarComponent, textField) => {
    //Solo necesitamos estas zonas si usamoslos
    handleChange({ name: "isCalendarNecessary", value: calendarComponent });
    handleChange({ name: "isInputNecessary", value: textField });
  };
  const handleSetFilter = (filterType) => {
    switch (filterType) {
      case "Sort By Date":
        handleExtraComponents(false, false);
        resetFilters();
        //Y llamamos nuestra mascara para dispatch(sortByDate())
        putSortDateFilter();
        break;
      case "Sort By Amount":
        handleExtraComponents(false, false);
        resetFilters();
        putAmountFilter();
        break;
      case "Put Text Filter":
        resetFilters();
        handleExtraComponents(false, true);
        break;
      case "Put Date Filter":
        resetFilters();
        handleExtraComponents(true, false);
        break;
      default:
    }
  };

  //Necesario para react componente de airbnb
  const onFocusChange = (focusedInput) => {
    handleChange({
      name: "focusedInput",
      value: !focusedInput ? "startDate" : focusedInput
    });
  };

  //Necesario para react componente de airbnb
  const onDatesChange = ({ startDate, endDate }) => {
    handleChange({ name: "startDate", value: startDate });
    handleChange({ name: "endDate", value: endDate });
    startDate && putStartDateFilter(startDate.valueOf());
    endDate && putEndDateFilter(endDate.valueOf());
  };

  return (
    <div>
      <select
        //Tiene los atributos mismos de input
        value={filterValue}
        onChange={(e) => handleSetFilter(e.target.value)}
      >
        <option>Sort By Date</option>
        <option>Sort By Amount</option>
        <option>Put Text Filter</option>
        <option>Put Date Filter</option>
      </select>

      {expenseFilterController.isInputNecessary && (
        <input
          type="text"
          value={filterValue}
          onChange={(e) => putTextFilter(e.target.value)}
        />
      )}
      {/* Puedes encontrar las docs sobre este componente en github o site de airbnb*/}
      {expenseFilterController.isCalendarNecessary && (
        <DayPickerRangeController
          startDate={expenseFilterController.startDate}
          endDate={expenseFilterController.endDate}
          onDatesChange={onDatesChange}
          focusedInput={expenseFilterController.focusedInput}
          onFocusChange={onFocusChange}
          initialVisibleMonth={() => moment()}
      
        />
      )}
    </div>
  );
};

//Mapeamos algunos atributos para utilizar funciones directamente.
//react-redux toma dispatch directamente
const mapDispatchToProps = (dispatch) => {
  return {
    //Estas funciones son solamente algunos atajos para usar mas bien que escribir el objeto directamente
    putTextFilter: (filter) => dispatch(setTextFilter(filter)),
    putStartDateFilter: (date) => dispatch(setStartDate(date)),
    putEndDateFilter: (date) => dispatch(setEndDate(date)),
    putSortDateFilter: () => dispatch(sortByDate()),
    putAmountFilter: () => dispatch(sortByAmount())
  };
};

//Usamos valor de filtro en input
const mapsStateToProps = (state) => {
  return {
    filterValue: state.filter
  };
};
//El primer para obtener el estado de store, y segundo para obtener acciones. y Por ultimo envolvemos nuestro componente
export default connect(
  mapsStateToProps,
  mapDispatchToProps
)(ExpenseListFilters);
