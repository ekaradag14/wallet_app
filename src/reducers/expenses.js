//Si tenemos un valor defecto complejo estas valores defectos sera util
const expensesReducerDefaultState = [];
//Cambiamos estado del expenses en redux con funciones de acciones
const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      //AÃ±adimos un expense nuevo
      return [...state, action.expense];
    case 'REMOVE_EXPENSE':
      return state.filter((expense) => expense.id !== action.id);
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        //Actualizamos un objeto soltero que concuerda con nuestra id
        if (expense.id === action.id) {
          return {
            //Todas las actualizaciones sobrescriben el estado previo
            ...expense,
            ...action.updates,
          };
        } else {
          return expense;
        }
      });

    default:
      return state;
  }
};

export default expensesReducer;
