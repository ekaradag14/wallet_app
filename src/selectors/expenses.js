const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((expense) => {
      //Usamos || por utilizar filtros aunque no tenemos todos los valores para filtros
      const startDateMatch =
        typeof startDate !== 'number' || expense.createdAt >= startDate;
      const endDateMatch =
        typeof endDate !== 'number' || expense.createdAt <= endDate;
      const textMatch =
        typeof text !== 'string' ||
        expense.description.toLowerCase().includes(text.toLocaleLowerCase());
      //Devolvemos un objeto si concuerdas a todos los filtros
      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      //sort funciÃ³n de javascript.
      if (sortBy === 'date') {
        // Ordenamos de nuevo a viejo
        return a.createdAt > b.createdAt ? 1 : -1; // En sort si queremos a ven antes de b devolvemos -1
      } else if (sortBy === 'amount') {
        return a.amount < b.amount ? 1 : -1;
      }
      return 0;
    });
};

export default getVisibleExpenses;
