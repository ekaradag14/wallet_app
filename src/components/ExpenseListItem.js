import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { removeExpense } from "../actions/expenses";

export const ExpenseListItem = ({
  description,
  amount,
  note,
  createdAt,
  id,
  remove
}) => {
  const handleRemoveExpense = () => {
    remove(id);
  };
  let editLink = `/edit/${id}`;
  return (
    <div>
      <h3> {description} </h3>
      <p>
        Your {description} was worth {amount} â‚º.
      </p>
      {note.length === 0 ? <p>You did not take any notes</p> : <p>Your noted {note} .</p>}
      <p> You made this expense at {new Date(createdAt).toDateString()}</p>
      <Link className="edit-todo" to={editLink}>
        Edit
      </Link>
      <button className="remove-todo" onClick={handleRemoveExpense}>
        Remove
      </button>
    </div>
  );
};

const mapDispatchToProps = (dispatch) => {
  return {
    remove: (id) => dispatch(removeExpense({ id }))
  };
};

export default connect(null, mapDispatchToProps)(ExpenseListItem);
