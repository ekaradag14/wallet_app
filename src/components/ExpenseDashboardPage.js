import React from "react";
//Usamos react-redux para alcanzar al store directamente de los componentes
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
//Importamos el función para filtrar nuestra store
import selectExpenses from "../selectors/expenses";
import ExpenseListFilters from "./ExpenseListFilters";

export const ExpenseDashboardPage = ({ expenses }) => { //Exportamos este para probar con jest.js
  const listIt = expenses.map((element) => (
    //Doy todo objeto en un manera rested
    <ExpenseListItem key={element.id} {...element} />
  ));

  return (
    <div>
      This is my expense list
      <ExpenseListFilters />
      {expenses.length === 0 ? <p>I have no expenses :)</p> : listIt }
    </div>
  );
};

const mapsStateToProps = (state) => {
  //Volvemos que necesitamos de store
  return {
    //Solo damos un estado filtrado para mostrar.
    expenses: selectExpenses(state.expenses, state.filters)
  };
  //Ahora tenemos gastos y filtros en props de ExpenseList
};

// El argumento primer es para función que desestructura el store. El argumento secundario es para a cual componente pasamos los partes de store.

export default connect(mapsStateToProps)(ExpenseDashboardPage);
