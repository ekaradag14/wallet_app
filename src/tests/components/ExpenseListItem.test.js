import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseListItem } from '../../components/ExpenseListItem';
import  expenses from '../fixtures/expenses';
import toJSON from 'enzyme-to-json';

describe('Expense List Item Test Scenarios', () => {
    test('should render the item correctly', () => {
        const wrapper = shallow(<ExpenseListItem {...expenses[0]} />)
        expect(toJSON(wrapper)).toMatchSnapshot();
    })

    test('should show default note text', () => {
        const wrapper = shallow(<ExpenseListItem {...expenses[2]} />)
         expect(toJSON(wrapper)).toMatchSnapshot();
    })
})