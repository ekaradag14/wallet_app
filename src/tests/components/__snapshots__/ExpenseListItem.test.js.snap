// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expense List Item Test Scenarios should render the item correctly 1`] = `
<div>
  <h3>
     
    Rent
     
  </h3>
  <p>
    Your 
    Rent
     was worth 
    220
     ₺.
  </p>
  <p>
    Your noted 
    This month's rent
     .
  </p>
  <p>
     You made this expense at 
    Thu Jan 01 1970
  </p>
  <Link
    className="edit-todo"
    to="/edit/0"
  >
    Edit
  </Link>
  <button
    className="remove-todo"
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;

exports[`Expense List Item Test Scenarios should show default note text 1`] = `
<div>
  <h3>
     
    Electricity
     
  </h3>
  <p>
    Your 
    Electricity
     was worth 
    150
     ₺.
  </p>
  <p>
    You did not take any notes
  </p>
  <p>
     You made this expense at 
    Thu Jan 01 1970
  </p>
  <Link
    className="edit-todo"
    to="/edit/2"
  >
    Edit
  </Link>
  <button
    className="remove-todo"
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
